storage/src/storage/sqlite/sync.rs

```diff
@@ -0,0 +1,923 @@
+            .query_row::<Vec<u8>, _, _>(r"SELECT hash FROM blocks WHERE canon_height = ?", [block_num], |row| {
+                row.get(0)
+            })
+            .optional()?
+            .map(|x| Digest::from(&x[..])))
+    }
+
+    fn get_block_header(&mut self, hash: &Digest) -> Result<SerialBlockHeader> {
+        self.conn
+            .query_row(
+                r"
+            SELECT
+                previous_block_hash,
+                merkle_root_hash,
+                pedersen_merkle_root_hash,
+                proof,
+                time,
+                difficulty_target,
+                nonce
+            FROM blocks WHERE hash = ?",
+                [hash],
+                |row| {
+                    Ok(SerialBlockHeader {
+                        previous_block_hash: row.get(0)?,
+                        merkle_root_hash: MerkleRootHash(read_static_blob(row, 1)?),
+                        pedersen_merkle_root_hash: PedersenMerkleRootHash(read_static_blob(row, 2)?),
+                        proof: ProofOfSuccinctWork(read_static_blob(row, 3)?),
+                        time: row.get(4)?,
+                        difficulty_target: row.get(5)?,
+                        nonce: row.get(6)?,
+                    })
+                },
+            )
+            .map_err(Into::into)
+    }
+
+    fn get_block_state(&mut self, hash: &Digest) -> Result<BlockStatus> {
+        let output: Option<Option<usize>> = self
+            .conn
+            .query_row(r"SELECT canon_height FROM blocks WHERE hash = ?", [hash], |row| {
+                row.get(0)
+            })
+            .optional()?;
+
+        Ok(match output {
+            None => BlockStatus::Unknown,
+            Some(None) => BlockStatus::Uncommitted,
+            Some(Some(n)) => BlockStatus::Committed(n),
+        })
+    }
+
+    fn get_block_states(&mut self, hashes: &[Digest]) -> Result<Vec<BlockStatus>> {
+        let mut out = Vec::with_capacity(hashes.len());
+        for hash in hashes {
+            let state = self.get_block_state(hash)?;
+            out.push(state);
+        }
+        Ok(out)
+        // let dyn_hashes: Vec<&dyn ToSql> = hashes.iter().map(|x| -> &dyn ToSql { x }).collect::<Vec<_>>();
+        // let mut stmt = self.conn.prepare_cached("SELECT hash, canon_height FROM blocks WHERE hash IN rarray(?)")?;
+        // let mut output = IndexMap::<Digest, BlockStatus>::new();
+        // for hash in hashes {
+        //     output.insert(hash.clone(), BlockStatus::Unknown);
+        // }
+        // for row in stmt.query_map(&dyn_hashes[..], |row| Ok((row.get(0)?, row.get(1)?)))? {
+        //     let (hash, canon_height): (Digest, Option<usize>) = row?;
+        //     *output.get_mut(&hash).unwrap() = match canon_height {
+        //         Some(height) => BlockStatus::Committed(height),
+        //         None => BlockStatus::Uncommitted,
+        //     };
+        // }
+        // Ok(output.into_iter().map(|x| x.1).collect())
```

Comment:
was this a potential optimization? are you still working on it or should it be removed?

Reply:
it was a superior implementation that didn't work. forgot the specifics.